internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,349 @@
+
+	// promote secondary to primary or force promote image if its already
+	// primary
+	if !mirroringInfo.Primary || force && mirroringInfo.Primary {
+		err = rbdVol.promoteImage(ctx, force)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to promote mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	return &replication.PromoteVolumeResponse{}, nil
+}
+
+// DemoteVolume demotes the volume as secondary.
+func (cs *ControllerServer) DemoteVolume(ctx context.Context,
+	req *replication.DemoteVolumeRequest,
+) (*replication.DemoteVolumeResponse, error) {
+	var rbdVol = &rbdVolume{}
+	var err error
+	// ger rbdVolume from VolumeID
+	rbdVol, err = cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol)
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		return nil, status.Errorf(codes.InvalidArgument, "image is in %d Mode", mirroringInfo.State)
+	}
+
+	// demote image to secondary
+	if mirroringInfo.Primary {
+		err = rbdVol.demoteImage(ctx)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to demote mirroring of %s with error %s", err.Error())
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	return &replication.DemoteVolumeResponse{}, nil
+}
+
+// ResyncVolume resync the volume.
+func (cs *ControllerServer) ResyncVolume(ctx context.Context,
+	req *replication.ResyncVolumeRequest,
+) (*replication.ResyncVolumeResponse, error) {
+	var rbdVol = &rbdVolume{}
+	var err error
+	// ger rbdVolume from VolumeID
+	rbdVol, err = cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol)
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		// TODO check if the image is not found return valid error message
+		// in case of Resync the image will get deleted and gets recreated and
+		// it takes time for this operation
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		return nil, status.Error(codes.InvalidArgument, "image mirroring is not enabled")
+	}
+
+	// return error if the image is still primary
+	if mirroringInfo.Primary {
+		return nil, status.Error(codes.InvalidArgument, "image is in primary state")
+	}
+
+	err = rbdVol.resyncImage(ctx)
+	if err != nil {
+		util.ErrorLog(ctx, "failed to resync %s with error %s", err.Error())
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	// TODO check the image state and return its ready to use or not
+	mirrorStatus, err := rbdVol.getImageMirroingStatus()
```

Comment:
Once we hit resync, isn't there a brief point in time where the image is recreated, IOW missing on the pool for some time till initial image metadata is recreated? The subsequent call may hence get a false negative if so.

Reply:
yes with some delay the image gets recreated after issuing the resync command. added check for the image not found error to return the image as not ready.

Reply:
Just checked NodeStage for any issues when image is not found in this state, and it looks clean. Other requests here would return an error till then and ready state would not be disturbed/updated till resync returns the same, so looks fine.

If ready is set, does that mean initial sync is complete and subsequent incremental sync is in progress? IOW, up+syncing denotes initial rebuild for resync and up+replaying/stopped denotes further snapshot sync?

Asking, as in Ramen we would want to wait for initial full resync to be complete before calling a secondary healthy and ready for any failback/over, and hence the ready response (and subsequently the VolumeReplication CR status) is useful to determine the same.

Reply:
Not sure about this one I need to test it out. if rbd allows mapping of the image whose state is up+syncing we can set ready to true. will check with @dillaman.